[5X] pg_upgrade_support: Detect views with deprecated tables

soumyadeep2007 opened the pull request:
This commit introduces a UDF and view query-tree walker to check for
deprecated tables referenced inside views.

deprecated_objects.c contains a list of deprecated tables and helpers
to check for deprecated tables. It will contain more types of
deprecated objects in the future.

The list of deprecated tables was obtained as follows:

1. Connected to template1 on a 6X cluster and ran:
```sql
COPY (SELECT relname FROM pg_class) TO '/tmp/6x_tables.csv' WITH CSV;
```

2. Connected to template1 on a 5X cluster and ran:
```sql
CREATE TEMP TABLE six_tables(a text);
COPY six_tables FROM '/tmp/6x_tables.csv';

SELECT oid AS deprecated_relation_oid, relname AS deprecated_relation_name
FROM pg_class
WHERE relkind = 'r' AND
relname IN
((select relname from pg_class) EXCEPT (select * from six_tables));
```

soumyadeep2007 commented:
Associated PR with tests: https://github.com/greenplum-db/gpupgrade/pull/552

dgkimura commented on contrib/pg_upgrade_support/deprecated_objects.c:
```diff
@@ -0,0 +1,94 @@
+
+static bool is_deprecated_table(Oid reloid);
+
+static const Oid deprecated_tables[] = {
```
I realize the commit message explains how this list was generated, but it may also be worth adding a comment too.

soumyadeep2007 replied:
Yup will do!

dgkimura commented on contrib/pg_upgrade_support/deprecated_objects.c:
```diff
@@ -0,0 +1,94 @@
+static bool is_deprecated_table(Oid reloid);
+
+static const Oid deprecated_tables[] = {
+	11937,	/* gp_toolkit.__gp_localid */
+	11939,	/* gp_toolkit.__gp_masterid */
```
Naive question, how do we keep these 2 `gp_toolkit.__gp_*` OIDs static?   I didn't realize they were hardcoded somewhere.

ashwinstar replied:
I highly doubt OIDs of `gp_toolkit` are static. One can even drop and recreate gp_toolkit schema.

soumyadeep2007 replied:
Oh wow, why then do they show up in `template1`?! That is why I thought they were truly part of the catalog.

Ugh, I think I found the answer:

```
/*
 * Load GPDB additions to the schema.
 *
 * These are contained in directory "cdb_init.d". We load all .sql files
 * from that directory, in alphabetical order. This modular design allows
 * extensions to put their install scripts under cdb_init.d, and have them
 * automatically installed directly in the template databases of every new
 * cluster.
 */
static void
setup_cdb_schema(void)
```

Well, we will have to come up with a different detection mechanism then for `gp_toolkit`. I don't have any ideas off the top of my head, so I'm going to remove these for now.

soumyadeep2007 replied:
Decided to dynamically sense the two Oids from gp_toolkit with a syscache lookup.

dgkimura reviewed (commented):
Looks good to me.

soumyadeep2007 commented on contrib/pg_upgrade_support/deprecated_objects.c:
```diff
@@ -0,0 +1,94 @@
+	if (IsA(node, RangeTblEntry))
+	{
+		RangeTblEntry *rte = (RangeTblEntry *) node;
+		return is_deprecated_table(rte->relid);
```
To handle subqueries and also CTEs, we use the recursion that we get with `query_tree_walker` to recurse into a subquery or a CTE's subquery.
The lines below handle [CTEs](https://github.com/greenplum-db/gpdb/pull/12072/files#diff-17a67be8f570ef9ca5c3275b90bc73084ec1dc0b7375228522bf2f69d0db345fR57-R80) and [subqueries](https://github.com/greenplum-db/gpdb/pull/12072/files#diff-17a67be8f570ef9ca5c3275b90bc73084ec1dc0b7375228522bf2f69d0db345fR81-R91)

Additionally, FYR this PR adds some tests which exercise CTEs, subqueries and joins.

soumyadeep2007 commented on contrib/pg_upgrade_support/pg_upgrade_support.c:
```diff
@@ -791,3 +797,30 @@ check_node_unknown_walker(Node *node, void *context)
+	Query		*viewquery;
+	bool		found;
+
+	if (rel == NULL)
```
Oh yes! Thanks for that!

soumyadeep2007 commented:
Thanks all for all the reviews, really appreciate it!

I addressed most of the feedback and this PR is ready for another round of review. Please take a look when you get a chance.
